dependency/retrieval/main.go

```
@@ -0,0 +1,134 @@
+	body, err := getAsString("https://www.python.org/downloads/")
+	if err != nil {
+		return nil, fmt.Errorf("could not retrieve new versions from upstream: %w", err)
+	}
+
+	re := regexp.MustCompile(`release-number.*Python ([\d]+\.[\d]+\.[\d]+)`)
+
+	var versions []string
+	for _, line := range strings.Split(body, "\n") {
+		matches := re.FindStringSubmatch(line)
+		if len(matches) == 2 {
+			versions = append(versions, matches[1])
+		}
+	}
+
+	return versionology.NewSimpleVersionFetcherArray(versions...)
+}
+
+func generateMetadata(versionFetcher versionology.VersionFetcher) ([]versionology.Dependency, error) {
+	version := versionFetcher.Version().String()
+	sourceURL := fmt.Sprintf("https://www.python.org/ftp/python/%[1]s/Python-%[1]s.tgz", version)
+
+	sourceSHA256, err := getSha256(sourceURL, version)
+	if err != nil {
+		return nil, err
+	}
+
+	configMetadataDependency := cargo.ConfigMetadataDependency{
+		CPE:            fmt.Sprintf("cpe:2.3:a:python:python:%s:*:*:*:*:*:*:*", version),
+		ID:             "python",
+		Licenses:       retrieve.LookupLicenses(sourceURL, upstream.DefaultDecompress),
+		Name:           "Python",
+		PURL:           retrieve.GeneratePURL("python", version, sourceSHA256, sourceURL),
+		Source:         sourceURL,
+		SourceChecksum: fmt.Sprintf("sha256:%s", sourceSHA256),
+		Version:        version,
+	}
+
+	return collections.TransformFuncWithError(supportedStacks, func(pair StackAndTargetPair) (versionology.Dependency, error) {
+		configMetadataDependency.Stacks = pair.stacks
+
+		if slices.Contains(pair.stacks, "*") {
+			configMetadataDependency.Checksum = configMetadataDependency.SourceChecksum
+			configMetadataDependency.URI = configMetadataDependency.Source
+		}
+		return versionology.NewDependency(configMetadataDependency, pair.target)
+	})
+}
+
+func getSha256(sourceURL string, version string) (string, error) {
+	resp, err := http.Get(sourceURL)
+	if err != nil {
+		panic(fmt.Errorf("failed to query url: %w", err))
```

If we are able to return the error (and do elsewhere in this function), why do we `panic` here?

---

Let's return the error.